src/backend/distributed/planner/distributed_planner.c

```diff
@@ -594,12 +717,14 @@ CreateDistributedPlannedStmt(uint64 planId, PlannedStmt *localPlan, Query *origi
 static DistributedPlan *
 CreateDistributedPlan(uint64 planId, Query *originalQuery, Query *query, ParamListInfo
```

Comment:
Thinking on this more:

There might be some tricky edge cases of not creating distributed plans and not using distributed execution (in this patch and for my above proposal as well), such as the following:

```SQL
BEGIN;

-- does a Postgres planning/execution, skipping Citus altogether
-- !!!! When asserts enabled, this crashes !!!!
SELECT count(*) FROM ref_table JOIN local_table ON (true);

-- does a distributed planning, which yields a distributed execution (e.g., connect to itself for the coordinator as well)
-- so a self deadlock
TRUNCATE ref_table;
```

Also, since we're skipping the distributed planning/execution, we might not be acquiring the correct shard locks etc. Even, `citus.log_remote_commands` wouldn't should show the query. Or, anything that Citus planner/executors does, that is relevant to the reference tables would be skipped. 

So, maybe we should re-architect the solution such that Citus is always in the picture, just conveying the plan to the local execution. In that case, we wouldn't need to consider one more edge case of queries on reference tables skipping the distributed planning/execution.

To be more specific, what do you think of treating local/ref joins as router plans? In that case, we can avoid any rough edges with this new change, (and later focus on improving the local execution). 

I guess, it shouldn't be too hard to refactor router planner to accept local tables. I think I'm even OK to have a new case, similar to `FastPathRouterPlanner()`, where we have `ReferenceTableLocalTableJoinRouterPlanner()`, which does the trick in `RouterPlanner()`.

I'm curious of your thoughts (and also @marcocitus if he's around).

Reply:
Overall this makes sense. I am splitting this PR into two, one which contains everything except planner related changes (https://github.com/citusdata/citus/pull/3155), and second will be planner related changes. I am sending a calendar invite for tomorrow to chat about planner changes.